### ISSUE TITLE
function expression as class prop initializer

### ISSUE THREAD

**Issue by aervin** [fimbullinter/wotan#232]
I'm wondering what would be the appropriate use-case for this code:
```ts
class Foo {
    public buzz = function () {
        this.mayExist.doSomething();
    };
}
```
versus this code:
```ts
class Bar {
    // must declare mustProp; obvious what "this" is in fizz()
    private mustProp: Foo;

    public fizz() {
        this.mustProp.doSomething();
    };
}
```

In the first example, it's confusing to me what `this` is in the anonymous function being assigned to `buzz`. TS defaults to `this: any`, which seems error-prone to me. 

Is the first example something worth linting against, or are there occasions when that approach makes sense?

**Comment by ajafff** [comment]
> can function declarations within the method complicate this approach?

You need to stop recursing on nodes that have their own `this` reference (functions, classes, methods, accessors) -> `hasOwnThisReference` from `tsutils`.
Otherwise the rule might have false negatives.
